src/main/java/javaslang/WrappedString.java

```diff
@@ -0,0 +1,2198 @@
+            left.addAll(right);
+            return left;
+        };
+        final Function<ArrayList<Character>, WrappedString> finisher = WrappedString::ofAll;
+        return Collector.of(supplier, accumulator, combiner, finisher);
+    }
+
+    /**
+     * Returns a singleton {@code String}, i.e. a {@code String} of one element.
+     *
+     * @param element An element.
+     * @return A new {@code String} instance containing the given element
+     */
+    public static WrappedString of(Character element) {
+        return new WrappedString(new java.lang.String(new char[] {element}));
+    }
+
+    /**
+     * Creates a String of the given elements.
+     *
+     * @param elements Zero or more elements.
+     * @return A string containing the given elements in the same order.
+     * @throws NullPointerException if {@code elements} is null
+     */
+    public static WrappedString of(Character... elements) {
+        Objects.requireNonNull(elements, "elements is null");
+        final char[] chrs = new char[elements.length];
+        for (int i = 0; i < elements.length; i++) {
+            chrs[i] = elements[i];
+        }
+        return new WrappedString(new java.lang.String(chrs));
+    }
+
+    /**
+     * Creates a String of {@code java.lang.String}.
+     *
+     * @param javaString {@code java.lang.String} instance.
+     * @return A new {@code javaslang.String}
+     */
+    public static WrappedString of(java.lang.String javaString) {
+        return javaString.isEmpty() ? empty() : new WrappedString(javaString);
+    }
+
+    /**
+     * Creates a String of the given elements.
+     *
+     * The resulting string has the same iteration order as the given iterable of elements
+     * if the iteration order of the elements is stable.
+     *
+     * @param elements An Iterable of elements.
+     * @return A string containing the given elements in the same order.
+     * @throws NullPointerException if {@code elements} is null
+     */
+    public static WrappedString ofAll(Iterable<? extends Character> elements) {
+        final StringBuilder sb = new StringBuilder();
+        for (Character character : elements) {
+            sb.append(character);
+        }
+        return sb.length() == 0 ? EMPTY : of(sb.toString());
+    }
+
+    //
+    //
+    // IndexedSeq
+    //
+    //
+
+    @Override
+    public WrappedString append(Character element) {
+        return of(back + element);
+    }
+
+    @Override
+    public WrappedString appendAll(Iterable<? extends Character> elements) {
+        Objects.requireNonNull(elements, "elements is null");
+        final StringBuilder sb = new StringBuilder(back);
+        for (char element : elements) {
+            sb.append(element);
+        }
+        return of(sb.toString());
+    }
+
+    @Override
+    public WrappedString clear() {
+        return EMPTY;
+    }
+
+    @Override
+    public Vector<Tuple2<Character, Character>> crossProduct() {
```

Comment:
I wanted to think about the usage of Vector. This is what I've come to:

Scala uses for `WrappedString.combinations(int)` as return type `Iterator<WrappedString>`. I also prefer not to use `Vector` _directly_. Instead I would use `Seq` or `Iterator`. `Seq` is not a good alternative because of the wildcards in return types, e.g. `Tuple2<? extends Seq, ...>`. This leads to `Iterator` in the case of this method (which could be obtained by calling `result.iterator()`).

In general there is another thing which comes to my mind. Instead of wrapping a sequence of chars, i.e. `java.lang.String`, we could wrap a sequence of arbitrary object, for example stored in `Vector<?> back`. This would `flatMap` allow to return a `WrappedString`. When a concrete String representation is needed, `back.mkString()` could be called. Example: `replaceAll(regEx, Str) = back.mkString().replaceAll(regEx, Str)`.

This is just an initial thought. Before going further I've to think more about it (later today).

@ruslansennov @patrox What are your thoughts?